plugins/destination/sqlite/client/migrate.go

```diff
@@ -38,70 +39,192 @@ func (i *tableInfo) getColumn(name string) *columnInfo {
-				return err
-			}
+
+		if !(change1.new && change1.newPk) && change2.new && change2.newPk {
+			return false
 		}
-		if err := c.Migrate(ctx, table.Relations); err != nil {
-			return err
+
+		return change1.name < change2.name
+	})
+
+	return columnChanges, nil
+}
+
+func (c *Client) getTableChange(ctx context.Context, table *schema.Table) (*tableChange, error) {
+	tableExist, err := c.isTableExistSQL(ctx, table.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if table %s exists: %w", table.Name, err)
+	}
+	tableChange := &tableChange{table: table, new: !tableExist}
+	if tableExist {
+		columnChanges, err := c.getColumnChanges(table)
+		if err != nil {
+			return nil, err
 		}
+		tableChange.columnChanges = columnChanges
 	}
-	return nil
+	return tableChange, nil
 }
 
-func (c *Client) isTableExistSQL(_ context.Context, table string) (bool, error) {
-	var tableExist int
-	if err := c.db.QueryRow(isTableExistSQL, table).Scan(&tableExist); err != nil {
-		return false, fmt.Errorf("failed to check if table %s exists: %w", table, err)
+func (c *Client) getSchemaChanges(ctx context.Context, tables schema.Tables) ([]*tableChange, error) {
+	changes := make([]*tableChange, len(tables))
+	for i, table := range tables {
+		tableChange, err := c.getTableChange(ctx, table)
+		if err != nil {
+			return nil, err
+		}
+		changes[i] = tableChange
+		for _, relation := range table.Relations {
+			relationChanges, err := c.getTableChange(ctx, relation)
+			if err != nil {
+				return nil, err
+			}
+			changes = append(changes, relationChanges)
+		}
 	}
-	return tableExist == 1, nil
+	return changes, nil
 }
 
-func (c *Client) autoMigrateTable(_ context.Context, table *schema.Table) error {
-	var err error
-	var info *tableInfo
+func getMigrationErrors(changes []*tableChange) []string {
+	var errors []string
+	for _, tableChange := range changes {
+		if tableChange.new {
+			continue
+		}
+		for _, colChange := range tableChange.columnChanges {
+			if colChange.new && colChange.newPk {
+				errors = append(errors, fmt.Sprintf("can't migrate table %q since adding the new PK column %q is not supported. Try dropping this table", tableChange.table.Name, colChange.name))
+				// no need to report other errors as the user needs to drop the table altogether
+				break
+			}
+			if !colChange.new && colChange.oldType != colChange.newType {
+				errors = append(errors, fmt.Sprintf("can't migrate table %q since changing the type of column %q from %q to %q is not supported. Try dropping this column for this table", tableChange.table.Name, colChange.name, colChange.oldType, colChange.newType))
+			}
+		}
+	}
+	return errors
+}
 
-	if info, err = c.getTableInfo(table.Name); err != nil {
-		return fmt.Errorf("failed to get table %s columns types: %w", table.Name, err)
+// This is the responsibility of the CLI of the client to lock before running migration
+func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
```

Comment:
We could move this part to the SDK and pass in functions to drop/add columns/tables if we need to